recipes/osx-sysroot/meta.yaml

```diff
@@ -0,0 +1,71 @@
+{% set name = "osx-sysroot" %}
+{% set version = "1" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+build:
+  number: 0
+
+outputs:
+  - name: macosx-sdk
+    version: {{ _MACOSX_SDK_VERSION_ }}
+    script: build_sdk.sh
+    build:
+      noarch: generic
+      # weigh down newer sdks
+      track_features:                                 # [is_default_sdk == "no"]
+        - macosx-sdk_{{ _MACOSX_SDK_VERSION_ }}       # [is_default_sdk == "no"]
+    requirements:
+      run_constrained:
+        # we cannot build for the future!
+        - sysroot_osx-64 <={{ _MACOSX_SDK_VERSION_ }}
+        - sysroot_osx-arm64 <={{ _MACOSX_SDK_VERSION_ }}
+    test:
+      commands:
+        - "[[ ${MACOSX_SDK_VERSION} == '{{ _MACOSX_SDK_VERSION_ }}' ]]"
+        - test -d "${MACOSX_SDK_CACHE_DIR}/MacOSX${MACOSX_SDK_VERSION}.sdk"
+        - test -d "${SDKROOT}"
+    about:
+      home: https://github.com/conda-forge/osx-sysroot-feedstock
+      summary: package to download MacOSX SDKs
+      license: BSD-3-Clause
+      license_family: BSD
+      license_file: LICENSE
+
+  - name: sysroot_{{ cross_target_platform }}
+    version: {{ _MACOSX_DEPLOYMENT_TARGET_ }}
+    script: build_sysroot.sh
+    build:
+      noarch: generic
+      run_exports:
+        strong:
+          - __osx >={{ _MACOSX_DEPLOYMENT_TARGET_ }}
+    requirements:
+      run:
+        # we can use newer SDKs to build for older systems
+        - macosx-sdk >={{ _MACOSX_DEPLOYMENT_TARGET_ }}
```

Comment:
It would be impossible to enforce a version to be used in conda-build recipes.

Reply:
No I don't think so. As long as all of the non-default versions have track features, then we always get the one without a feature.

Reply:
How do you enforce https://github.com/conda-forge/python-feedstock/blob/main/.ci_support/osx_64_build_typedebugchannel_targetsconda-forge_python_debug.yaml#L1-L4 ?

Reply:
If you want a specific SDK, you put it in build. Otherwise, the solver picks the one with the fewest features. This will be 10.13 for intel and 11.0 for arm.

Reply:
While trying to remove an implicit feature, you are adding another implicit feature. Please let's not do that.

Reply:
Sure. If 

 - we are not happy with implicit features
 - we don't want to enforce that the SDK version is always equal to the deployment target 
 - we don't want to always default to the newest SDK

then we have to require that every user explicitly list both the sdk version and the deployment target in build. In this case, I would suggest removing cross target platform to make life easier for folks or having the jinja2 function take care of it somehow.

FWIW, I am not so unhappy with the implicit feature here. It arises because the deployment target and SDK can have different versions, and because defaulting to the newest SDK is not our standard policy. 

For sure once we add some special `{{ sysroot('default') }}` jinja2 magic, we won't need the feature on linux anymore.

Reply:
I also don't see the issue with defaulting to SDK == deployment target. Most feedstocks do not care about the distinction, and in effect, we've been doing just that with having both at 10.9.

Having an opt-in to change the SDK sounds nice, in the sense that few feedstocks will have to concern themselves with this distinction (but those that must, can do it).

Reply:
SDK is not needed for any metadata. Only the deployment target is needed for `__osx`. I don't see a reason to add this added complication that the user needs to modify in the recipe itself. Currently, to change the SDK version, you only to edit cbc.yaml, whereas with your suggestion we need to do both cbc.yaml and the recipe. The recipe's requirements does not benefit at all from the SDK version.

Reply:
So to summarize, we could:
* keep `MACOSX_SDK_VERSION` purely in CBC (global, with per-feedstock override)
* have `c_stdlib_name` and `c_stdlib_version` populated with sysroot (linux) and macosx_deployment_target (osx)

A recipe would then only need
```
- {{ stdlib("c") }}
```
and potentially overriding any of the following in the feedstock CBC:
```
MACOSX_SDK_VERSION:             # [osx}
  - 11.0                        # [osx]

c_stdlib_name:                  # [unix]
  - sysroot                     # [linux]
  - macosx_deployment_target    # [osx]
c_stdlib_version:               # [unix]
  - 2.12                        # [linux]
  - 10.13                       # [osx]
```

Reply:
Right I forgot to mention. The way I was doing this, with the implicit feature version here, we can completely remove the OSX SDK stuff from the pinnings and only leave the default deployment in the pinnings. We can also remove the download stuff and handling in the conda-forge-ci-setup package.

Reply:
> The recipe's requirements does not benefit at all from the SDK version.

This doesn't seem to be true for the python example you sent in that the SDK version appears to matter there and so it's version being in the recipe build metadata is useful in defining the build environment.

Reply:
Fair enough, that'd be an improvement on top of solving the whole `__osx` thing. Avoiding the downloads on every CI run would be nice indeed!

> I don't see a reason to add this added complication that the user needs to modify in the recipe itself.

While the SDK is not necessary for the artefact metadata, it's still a pretty essential piece of information for being able to build the recipe at all (if it doesn't work with the default). I find things in `meta.yaml` more explicit than those tucked away in cbc, so personally I don't see it as such a big problem to specify the sdk there.

Reply:
The other thing is that if we don't ship an SDK package, then the user has to manually make sure the SDK version and deployment target in the recipe are compatible. Avoiding this is what motivated me to make the SDK package in the first place.

Reply:
Sounds consistent with pushing all that implicit metadata into packages to me.

Reply:
So one option that might be closer to what @isuruf is suggesting is to keep the SDK download outside of a package but have the deployment target package activation script check the its version is less than the SDK version. We could print an error explaining what to do.

This would make the build fail with a known error as opposed to w/e happens when you try to build in an unsupported configuration. Of course we'd be throwing an error when we could have fixed this for the user, but that might be an OK compromise here.

Reply:
no more features!